File: Script/WellCare.mjs
import fetch from 'node-fetch';
import { JSDOM } from 'jsdom';
import { promises as fs } from 'fs';
import admin from 'firebase-admin';

// Initialize Firebase
import serviceAccount from './wellCareKey.json' assert { type: "json" };

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});
const db = admin.firestore();

// Function to fetch request verification token
async function fetchRequestVerificationToken() {
  const response = await fetch('https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Quote/GetQuotePartial', {
    method: 'GET',
    headers: {
      'Accept': 'text/html',
      'X-Requested-With': 'XMLHttpRequest',
    }
  });

  const text = await response.text();
  const dom = new JSDOM(text);

  const token = dom.window.document.querySelector('input[name="__RequestVerificationToken"]').value;
  return token;
}

// Function to get plan data
async function getPlanData(zipCode, countyName, stateAbbreviation, fipsCode, token) {
  const stateFipsCode = fipsCode.slice(0, 2);

  const headers = {
    'Content-Type': 'application/x-www-form-urlencoded',
    'Accept': 'text/html',
    'X-Requested-With': 'XMLHttpRequest',
    'Cookie': '_gcl_au=1.1.2020630245.1719605110; invoca_session=%7B%22ttl%22%3A%222024-07-28T20%3A05%3A10.845Z%22%2C%22session%22%3A%7B%7D%2C%22config%22%3A%7B%22ce%22A true%2C%22fv%22A true%7D%7D; _fbp=fb.1.1719605111194.350584963271778211; _ce.irv=new; cebs=1; _ce.s=v~b74d33180398624120fc303fa287cffa5a3c0b26~lcw~1719605111211~lva~1719605111211~vpv~0~lcw~1719605111213; _gid=GA1.2.270489795.1719836108; _clck=1nid4l2%7C2%7Cfn3%7C0%7C1640; _ce.clock_data=16%2C38.122.14.76%2C1%2C10f9287deaf609ee36fb37783f2b89c0%2CChrome%2CUS; _ga_J79ERXWWW5=GS1.2.1719845309.4.0.1719845349.0.0.0; UseNewDesignFeature=False; .AspNetCore.Culture=c%3Den-US%7Cuic%3Den-US; AspNetCore_Session_Cookie=CfDJ8HAEUs1AUs1NvxqdTq8XexaAQzEubvRGw99geCS18kJIumocrYceAE8Q3%2Fz1G5xUUwyBKzZgv2rZhixhjsqeZjeKNiSdSdVaSvLGh4XEuFBhtiTOnadb%2B97AZAZQxNdt1ZYSacLoJ11HoYiXEKX27DvW8Ec56erSXPTFAIz9FpIR; .AspNetCore.Antiforgery.9Ldf4oLyS6k=CfDJ8HAEUs1AUs1NvxqdTq8XexZzvsiCSkqK28qygeBtFz914qBqE0CPxehwX3nlMzC6tF33QvbaiKVxaDLusSrgDOXFnHXzxwDJJ1MPzDuabAP_eouNoLmFVQv2ZQM9d-UTAVOjiWHeGE9IY22gNGm89Cc; _ga_2P7LX36Q85=GS1.1.1719857091.7.1.1719857743.0.0.0; _ga=GA1.2.2013862579.1719605110; _gat_UA-179700527-1=1; _uetsid=8e9ac66037a311ef84b6459da3dce797; _uetvid=b92f45b0358911ef9c6db55e7d33790b; _clsk=1iep2st%7C1719857749160%7C10%7C1%7Cf.clarity.ms%2Fcollect; _gat_UA-43803870-1=1; _ga_GDQD34846P=GS1.1.1719857092.6.1.1719857800.1.0.0; _ga_XNQEBXVC3J=GS1.1.1719857092.6.1.1719857800.1.0.0'
  };

  const payload = new URLSearchParams({
    'quote.Location.Zip': zipCode,
    'quote.Location.County': countyName,
    'quote.Location.State': stateAbbreviation,
    'quote.Location.CountyFipsCode': fipsCode,
    'quote.Location.StateFipsCode': stateFipsCode,
    '__RequestVerificationToken': token
  });

  try {
    const response = await fetch(
      'https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Quote/GetQuotePartial',
      {
        method: 'POST',
        headers: headers,
        body: payload.toString()
      }
    );

    const text = await response.text();
    const dom = new JSDOM(text);

    const scriptTags = dom.window.document.querySelectorAll('script');
    let plansJsonString = null;

    scriptTags.forEach(script => {
      if (script.textContent.includes('var plans =')) {
        plansJsonString = script.textContent.match(/var plans = (\[.*?\]);/)[1];
      }
    });

    if (!plansJsonString) {
      console.log(`No plans data found in response for Zip Code: ${zipCode}`);
      return [];
    }

    const plans = JSON.parse(plansJsonString);
    const planElements = dom.window.document.querySelectorAll('.list-col');
    const planDetails = Array.from(planElements).map(el => {
      const name = el.querySelector('h3')?.textContent.trim();
      const contractId = el.querySelector('.section-block p')?.textContent.trim();
      return { name, contractId };
    });

    if (planDetails.length === 0) {
      console.log(`No plan details found in response for Zip Code: ${zipCode}`);
      return [];
    }

    const planData = planDetails.map(detail => {
      const plan = plans.find(p => `${p.Contract}-${p.PbpId}` === detail.contractId);
      if (!plan) {
        return null;
      }
      return {
        productContractCode: `${plan.Contract}-${plan.PbpId}-${plan.SegmentId}`,
        zipCode: zipCode,
        rating: plan.StarRating === 0 ? null : plan.StarRating,
        planName: detail.name,
        planID: plan.PlanId
      };
    }).filter(plan => plan !== null);

    if (planData.length === 0) {
      console.log(`No matching plans found for Zip Code: ${zipCode}`);
      return [];
    }

    for (const plan of planData) {
      try {
        const detailResponse = await fetch(`https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Plan/Index?PlanId=${plan.planID}&FormId=${plan.FormId}`, {
          method: 'GET',
          headers: headers
        });
        const detailText = await detailResponse.text();
        const detailDom = new JSDOM(detailText);

        const evidenceLinkElement = Array.from(detailDom.window.document.querySelectorAll('a')).find(a => a.textContent.includes('Evidence of Coverage'));
        if (evidenceLinkElement) {
          const evidenceLink = evidenceLinkElement.href;
          plan.evidenceOfCoverageLink = evidenceLink.startsWith('http') ? evidenceLink : `https://wellcare.isf.io${evidenceLink}`;
        } else {
          plan.evidenceOfCoverageLink = null;
        }
      } catch (detailError) {
        plan.evidenceOfCoverageLink = null;
      }
    }

    return planData;

  } catch (error) {
    console.error(`Error fetching plan data for Zip Code: ${zipCode}`, error);
    return [];
  }
}

// Function to save data to Firebase
async function saveToFirebase(plans) {
  const maxRetries = 5;

  const commitBatch = async (batch) => {
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        await batch.commit();
        break;
      } catch (e) {
        console.log(`Resource exhausted, retrying... ${e}`);
        await new Promise(res => setTimeout(res, 2 ** attempt * 1000)); // Exponential backoff
      }
    }
  };

  const batchSize = 10;

  // Save plans
  const plansCollection = db.collection('plans');
  for (let i = 0; i < plans.length; i += batchSize) {
    const batch = db.batch();
    const planChunk = plans.slice(i, i + batchSize);
    planChunk.forEach(plan => {
      const planRef = plansCollection.doc(`${plan.productContractCode}-${plan.zipCode}`);
      const planData = { ...plan };
      if (planData.rating === 0) {
        planData.rating = null;
      }
      batch.set(planRef, planData);
    });
    await commitBatch(batch);
    console.log(`Saved ${planChunk.length} plans to Firebase.`);
  }
}

async function main() {
  try {
    const zipMapping = JSON.parse(await fs.readFile('./final_zip_mapping.json', 'utf-8'));
    let allPlans = new Map();

    const token = await fetchRequestVerificationToken();
    const sortedZipCodes = Object.keys(zipMapping).sort((a, b) => a.localeCompare(b, undefined, { numeric: true }));

    for (const zipCode of sortedZipCodes) {
      console.log(`Checking ZIP code: ${zipCode}`);

      const { county_name, state_abbreviation, fips_codes } = zipMapping[zipCode];

      for (const fipsCode of fips_codes) {
        const plans = await getPlanData(zipCode, county_name, state_abbreviation, fipsCode, token);

        if (plans.length > 0) {
          plans.forEach(plan => {
            allPlans.set(`${plan.productContractCode}-${plan.zipCode}`, plan);
          });
        }

        if (allPlans.size >= 100) {
          await saveToFirebase(Array.from(allPlans.values()));
          allPlans.clear();
        }
      }
    }

    if (allPlans.size > 0) {
      await saveToFirebase(Array.from(allPlans.values()));
    }

    console.log('Plan information has been saved to Firebase');
  } catch (error) {
    console.error('Error:', error);
  }
}

// Run the main function
main();

File: Script/WellCare.py
import requests
from bs4 import BeautifulSoup
import json
import time
import firebase_admin
from firebase_admin import credentials, firestore
from urllib.parse import urlencode, urljoin

# Initialize Firebase
cred = credentials.Certificate('./wellCareKey.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

# Function to fetch a new cookie
def fetch_new_cookie():
    session = requests.Session()
    session.get('https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Quote/GetQuotePartial')
    return session.cookies.get_dict()

# Function to fetch request verification token
def fetch_request_verification_token(cookie):
    headers = {
        'Accept': 'text/html',
        'X-Requested-With': 'XMLHttpRequest',
        'Cookie': cookie
    }
    response = requests.get('https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Quote/GetQuotePartial', headers=headers)
    soup = BeautifulSoup(response.text, 'html.parser')
    token = soup.find('input', {'name': '__RequestVerificationToken'})['value']
    return token

# Initialize cookie and token
cookie_dict = fetch_new_cookie()
cookie = '; '.join([f'{key}={value}' for key, value in cookie_dict.items()])
token = fetch_request_verification_token(cookie)

# Function to get plan data
def get_plan_data(zip_code, county_name, state_abbreviation, fips_code):
    global cookie, token  # Make sure to use the updated cookie and token
    state_fips_code = fips_code[:2]

    headers = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Accept': 'text/html',
        'X-Requested-With': 'XMLHttpRequest',
        'Cookie': cookie
    }

    payload = {
        'quote.Location.Zip': zip_code,
        'quote.Location.County': county_name,
        'quote.Location.State': state_abbreviation,
        'quote.Location.CountyFipsCode': fips_code,
        'quote.Location.StateFipsCode': state_fips_code,
        '__RequestVerificationToken': token
    }

    response = requests.post('https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Quote/GetQuotePartial', headers=headers, data=urlencode(payload))
    
    # Refresh token and cookie if needed
    if response.status_code == 403:  # Forbidden, possibly due to expired token
        cookie_dict = fetch_new_cookie()
        cookie = '; '.join([f'{key}={value}' for key, value in cookie_dict.items()])
        token = fetch_request_verification_token(cookie)
        headers['Cookie'] = cookie
        payload['__RequestVerificationToken'] = token
        response = requests.post('https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Quote/GetQuotePartial', headers=headers, data=urlencode(payload))

    soup = BeautifulSoup(response.text, 'html.parser')

    script_tags = soup.find_all('script')
    plans_json_string = None

    for script in script_tags:
        if script.string and 'var plans =' in script.string:
            plans_json_string = script.string.split('var plans = ')[1].split(';')[0]

    if not plans_json_string:
        print(f"No plans data found in response for Zip Code: {zip_code}")
        return []

    plans = json.loads(plans_json_string)
    plan_elements = soup.select('.list-col')
    plan_details = [{'name': el.find('h3').text.strip(), 'contractId': el.select_one('.section-block p').text.strip()} for el in plan_elements]

    if not plan_details:
        print(f"No plan details found in response for Zip Code: {zip_code}")
        return []

    plan_data = [
        {
            'productContractCode': f"{plan['Contract']}-{plan['PbpId']}-{plan['SegmentId']}",
            'zipCode': zip_code,
            'rating': None if plan['StarRating'] == 0 else plan['StarRating'],
            'planName': detail['name'],
            'planID': plan['PlanId']
        }
        for detail in plan_details
        for plan in plans
        if f"{plan['Contract']}-{plan['PbpId']}" == detail['contractId']
    ]

    if not plan_data:
        print(f"No matching plans found for Zip Code: {zip_code}")
        return []

    for plan in plan_data:
        try:
            detail_response = requests.get(f"https://wellcare.isf.io/2024/g/f010acf28d0e468cb2aa59e0e68d694b/Plan/Index?PlanId={plan['planID']}", headers=headers)
            detail_soup = BeautifulSoup(detail_response.text, 'html.parser')
            evidence_link_element = next((a for a in detail_soup.find_all('a') if 'Evidence of Coverage' in a.text), None)
            if evidence_link_element:
                evidence_link = evidence_link_element['href']
                plan['evidenceOfCoverageLink'] = evidence_link if evidence_link.startswith('http') else urljoin('https://wellcare.isf.io', evidence_link)
            else:
                plan['evidenceOfCoverageLink'] = None
        except Exception as detail_error:
            print(f"Error fetching plan details for PlanID {plan['planID']}: {detail_error}")
            plan['evidenceOfCoverageLink'] = None

    return plan_data

# Function to save data to Firebase
def save_to_firebase(plans):
    max_retries = 5

    def commit_batch(batch):
        for attempt in range(max_retries):
            try:
                batch.commit()
                break
            except Exception as e:
                print(f"Resource exhausted, retrying... {e}")
                time.sleep(2 ** attempt)

    batch_size = 500  # Increased batch size to reduce the number of writes

    plans_collection = db.collection('WellCarePlans')
    for i in range(0, len(plans), batch_size):
        batch = db.batch()
        plan_chunk = plans[i:i + batch_size]
        for plan in plan_chunk:
            plan_ref = plans_collection.document(f"{plan['productContractCode']}-{plan['zipCode']}")
            batch.set(plan_ref, plan)
        commit_batch(batch)
        print(f"Saved {len(plan_chunk)} plans to Firebase.")

def main(start_zip_code=None):
    try:
        with open('./final_zip_mapping.json', 'r') as f:
            zip_mapping = json.load(f)
        all_plans = {}

        sorted_zip_codes = sorted(zip_mapping.keys(), key=lambda x: (len(x), x))

        if start_zip_code:
            if start_zip_code in sorted_zip_codes:
                start_index = sorted_zip_codes.index(start_zip_code)
                sorted_zip_codes = sorted_zip_codes[start_index:]
            else:
                print(f"Start ZIP code {start_zip_code} not found in the ZIP code mapping.")
                return

        for zip_code in sorted_zip_codes:
            print(f"Checking ZIP code: {zip_code}")

            county_name = zip_mapping[zip_code]['county_name']
            state_abbreviation = zip_mapping[zip_code]['state_abbreviation']
            fips_codes = zip_mapping[zip_code]['fips_codes']

            for fips_code in fips_codes:
                plans = get_plan_data(zip_code, county_name, state_abbreviation, fips_code)

                if plans:
                    for plan in plans:
                        all_plans[f"{plan['productContractCode']}-{plan['zipCode']}"] = plan

                if len(all_plans) >= 100:
                    save_to_firebase(list(all_plans.values()))
                    all_plans.clear()

        if all_plans:
            save_to_firebase(list(all_plans.values()))

        print('Plan information has been saved to Firebase')
    except Exception as error:
        print('Error:', error)

if __name__ == "__main__":
    start_zip = "15455"  # Specify the starting ZIP code here
    main(start_zip_code=start_zip)
